### PR TITLE
[CIFuzz] Download code coverage 

### PR DESCRIPTION
**Pull request by Leo-Neat**
This is the foundation for affected fuzzers. It provides the ability to map fuzzers to files. In the future we will use this functionality to pick which fuzzers to run during CI.

### CODE REVIEW

File: infra/cifuzz/cifuzz.py [hunk 1]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+  """
+  project_report_url = fuzz_target.url_join(fuzz_target.GCS_BASE_URL,
+                                            PROJECT_COVERAGE_REPORT,
+                                            project_name + '.json')
+  project_cov_json = get_json_from_url(project_report_url)
+  if not project_cov_json:
+    logging.error('Could not get the coverage report json from %s.',
+                  project_report_url)
+    return None
+  if 'fuzzer_stats_dir' not in project_cov_json:
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: I don't think this check belongs to this function, also I can hardly imagine us having such a JSON file uploaded

File: infra/cifuzz/cifuzz.py [hunk 2]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+    logging.error('fuzzer_stats_dir is not a key in downloaded json.')
+    return None
+  return project_cov_json
+
+
+def get_target_coverage_report(project_cov_report, target_name):
+  """Get the coverage report from a specific fuzz target.
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: super nit: "for a specific fuzz target"

File: infra/cifuzz/cifuzz.py [hunk 3]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+
+  Args:
+    project_cov_report: A dict containing the project cov report data.
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: IMO `latest_report_info` makes more sense. A report is a large JSON that has coverage counters, filepaths and other stuff. This is just a tiny metadata file, not a report

File: infra/cifuzz/cifuzz.py [hunk 4]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+    target_name: The name of the fuzz target whos coverage is requested.
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: nit: "whos" -> "whose" ?

File: infra/cifuzz/cifuzz.py [hunk 5]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+
+  Returns:
+    The targets coverage json dict or None on failure.
+  """
+  if 'fuzzer_stats_dir' not in project_cov_report:
+    return None
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: should we log anything in this case?

File: infra/cifuzz/cifuzz.py [hunk 6]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+
+  Returns:
+    A list of files that the fuzzer covers or None.
+  """
+  if not oss_fuzz_project_base:
+    return None
+  target_cov = get_target_coverage_report(project_cov_report, target_name)
+  if not target_cov:
+    return None
+  if not target_cov['data'][0]['files']:
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: assign this to a local variable e.g. `files` or `coverage_per_file`, and re-use it in the check plus on the line 283

File: infra/cifuzz/cifuzz.py [hunk 7]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+
+
+def get_files_covered_by_target(project_cov_report, target_name,
+                                oss_fuzz_project_base):
+  """Gets a list of files covered by the specific fuzz target.
+
+  Args:
+    project_cov_report: A dict containing the project cov report data.
+    target_name: The name of the fuzz target whose coverage is requested.
+    oss_fuzz_project_base: The location where OSS-Fuzz project is cloned too for
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: nit: "too" -> "to"

File: infra/cifuzz/cifuzz_test.py [hunk 8]
```diff
@@ -217,5 +218,104 @@ def test_parse_invalid_output(self):
...
+  def test_get_valid_project(self):
+    """Tests that a project's coverage report can be downloaded and parsed.
+
+    NOTE: This test relies on the test_project repo's coverage report.
+    Example was not used because it has no coverage reports.
+    """
+    cov_report = cifuzz.get_project_coverage_report(self.test_project)
+    self.assertIsNotNone(cov_report)
+    self.assertEqual(type(cov_report), dict)
+    self.assertTrue('fuzzer_stats_dir' in cov_report)
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz_test.py]: I don't think you need to test such details like these keys, just mock the download method and assert that it is being called with the correct expected URL

File: infra/cifuzz/cifuzz_test.py [hunk 9]
```diff
@@ -217,5 +218,104 @@ def test_parse_invalid_output(self):
...
+  """Test to get the files covered by a fuzz target in the cifuzz module."""
+
+  example_cov_json = 'example_curl_cov.json'
+  example_fuzzer = 'curl_fuzzer'
+  example_curl_file_list = 'example_curl_file_list'
+
+  def setUp(self):
+    with open(os.path.join(TEST_FILES_PATH, self.example_cov_json),
+              'r') as file:
+      self.envoy_cov_exmp = json.loads(file.read())
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz_test.py]: please use more descriptive variable names, e.g. `coverage_json_expected`

File: infra/cifuzz/cifuzz.py [hunk 10]
```diff
@@ -62,6 +65,9 @@
     'minidump has been written',
 ]
 
+# The path to get project's latest report json files.
+PROJECT_COVERAGE_REPORT = 'oss-fuzz-coverage/latest_report_info/'
```
[the remark this answers is not in the document]
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Changed.

File: infra/cifuzz/cifuzz.py [hunk 11]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+def get_project_coverage_report(project_name):
+  """Gets the report json for a specific OSS-Fuzz project from GCS.
+
+  Args:
+    project_name: The name of the relevant OSS-Fuzz project.
+
+  Returns:
+    The projects coverage json dict or None on failure.
+  """
+  project_report_url = fuzz_target.url_join(fuzz_target.GCS_BASE_URL,
```
[the remark this answers is not in the document]
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Added.

File: infra/cifuzz/cifuzz.py (hunk 1 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Removed.

File: infra/cifuzz/cifuzz.py (hunk 3 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: I changed it to `latest_cov_info`, does that work for you?

File: infra/cifuzz/cifuzz.py (hunk 2 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed.

File: infra/cifuzz/cifuzz.py (hunk 5 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Added.

File: infra/cifuzz/cifuzz.py (hunk 6 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Done.

File: infra/cifuzz/cifuzz.py (hunk 7 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed.

File: infra/cifuzz/cifuzz.py [hunk 12]
```diff
@@ -203,6 +209,109 @@ def run_fuzzers(fuzz_seconds, workspace, project_name):
...
+  """
+  try:
+    response = urllib.request.urlopen(url)
+  except urllib.error.HTTPError:
+    logging.error('HTTP error with url %s.', url)
+    return None
+  try:
+    result_json = json.loads(response.read().decode())
+  except ValueError as excp:
+    logging.error('Loading json failed with: %s.', str(excp))
```
[the remark this answers is not in the document]
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Done.

File: infra/cifuzz/cifuzz_test.py (hunk 9 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz_test.py]: Fixed.

File: infra/cifuzz/cifuzz_test.py (hunk 8 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz_test.py]: Fixed.

File: infra/cifuzz/cifuzz.py [hunk 13]
```diff
@@ -243,6 +248,110 @@ def check_fuzzer_build(out_dir):
...
+    logging.info('No files found in coverage report.')
+    return None
+
+  # Cases like curl there is /src/curl and /src/curl_fuzzers/ are handled.
+  if not oss_fuzz_project_base.endswith('/'):
+    oss_fuzz_project_base += '/'
+
+  affected_file_list = []
+  for file in coverage_per_file:
+    if file['filename'].startswith(
```
Comment (oliverchang) [inline infra/cifuzz/cifuzz.py]: nit:

consider formatting like this:
```python
if file['filename'].startswith(oss_fuzz_project_base) and
   file['summary']['regions']['count']:
```

File: infra/cifuzz/cifuzz.py [hunk 14]
```diff
@@ -243,6 +248,110 @@ def check_fuzzer_build(out_dir):
...
+  Args:
+    latest_cov_info: A dict containing a project's latest cov report info.
+    target_name: The name of the fuzz target whose coverage is requested.
+    oss_fuzz_project_base: The location where OSS-Fuzz project is cloned to for
+      the projects build.
+
+  Returns:
+    A list of files that the fuzzer covers or None.
+  """
+  if not oss_fuzz_project_base:
```
Comment (oliverchang) [inline infra/cifuzz/cifuzz.py]: When would it ever make sense for this condition to be true? Should we throw an exception instead?

File: infra/cifuzz/cifuzz.py [hunk 15]
```diff
@@ -243,6 +248,110 @@ def check_fuzzer_build(out_dir):
...
+
+  Args:
+    latest_cov_info: A dict containing a project's latest cov report info.
+    target_name: The name of the fuzz target whose coverage is requested.
+
+  Returns:
+    The targets coverage json dict or None on failure.
+  """
+  if 'fuzzer_stats_dir' not in latest_cov_info:
+    logging.error('The latest coverage report information did not cointain' \
```
[the remark this answers is not in the document]
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed. Thanks! I wasn't aware of that, I will keep it in mind for the future.

File: infra/cifuzz/cifuzz.py (hunk 14 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: You are right, It should never be true. An exception makes sense.

File: infra/cifuzz/cifuzz.py (hunk 13 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: I get a error with that formatting. It seems that would work If I added parentheses around the if statement.

File: infra/cifuzz/cifuzz.py [hunk 16]
```diff
@@ -243,6 +248,110 @@ def check_fuzzer_build(out_dir):
...
+  Returns:
+    Json dict or None on failure.
+  """
+  try:
+    response = urllib.request.urlopen(url)
+  except urllib.error.HTTPError:
+    logging.error('HTTP error with url %s.', url)
+    return None
+  try:
+    result_json = json.loads(response.read().decode())
```
[the remark this answers is not in the document]
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Fixed.

File: infra/cifuzz/cifuzz.py (hunk 13 above)
Comment (oliverchang) [reply infra/cifuzz/cifuzz.py]: My bad, yep you need the parentheses.

File: infra/cifuzz/cifuzz.py [hunk 17]
```diff
@@ -243,6 +249,113 @@ def check_fuzzer_build(out_dir):
...
+
+  # Cases like curl there is /src/curl and /src/curl_fuzzers/ are handled.
+  if not oss_fuzz_project_base.endswith('/'):
+    oss_fuzz_project_base += '/'
+
+  affected_file_list = []
+  for file in coverage_per_file:
+    if file['filename'].startswith(
+        oss_fuzz_project_base) and file['summary']['regions']['count']:
+      affected_file_list.append(file['filename'].replace(
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: nit: let's put the argument of this call into another variable, e.g. `relative_path = file['filename'].replace(...)` and then append it

File: infra/cifuzz/cifuzz.py [hunk 18]
```diff
@@ -243,6 +249,113 @@ def check_fuzzer_build(out_dir):
...
+    logging.info('No files found in coverage report.')
+    return None
+
+  # Cases like curl there is /src/curl and /src/curl_fuzzers/ are handled.
+  if not oss_fuzz_project_base.endswith('/'):
+    oss_fuzz_project_base += '/'
+
+  affected_file_list = []
+  for file in coverage_per_file:
+    if file['filename'].startswith(
```
Comment (Dor1s) [inline infra/cifuzz/cifuzz.py]: another potential readability improvement here would be:

```
if not file['filename'].startswith(oss_fuzz_project_base):
  continue

if file['summary']['regions']['count'] == 0:
  # Don't consider a file affected if code in it is never executed.
  continue

<replace and append logic here>
```

Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Added.

File: infra/cifuzz/cifuzz.py (hunk 17 above)
Comment (Leo-Neat) [reply infra/cifuzz/cifuzz.py]: Added.

**Pull request merged**